27 panda parsing

kgarwoodsdzwa opened the pull request:
Tools to attempt to isolate significant events in a wav file for unlabeled audio. Visualize the RMS and Mel-spectrogram for a wav file, or extract segments that exceed the average RMS for a wav file to inspect closer or label. Tested on one week deployment of passive acoustic recorders at panda site and was able to determine a repeated feeding time where panda chewing can be heard.

kgarwoodsdzwa committed:
add script to create 'loud' segments

needs some work to go through numerous sound files and give
better resulting filenames. but this script takes a sound file
and then calculates the rms across a certain number of frames you
tell it the size and hop for, giving you an array of rms values.
It then calculates the average rms, and multiplies it by 1.5
and creates 3 second segments of audio centered around the
frame that exceeded the rms threshold. it will not create
overlapping segments. there's definitely some potential issues
with this but for now seems to be able to create segments based
on the relative loudness of the whole sample. for example it
hasn't had to handle creating a segment where the values that
exceed the threshold equate for more than the specified 3s for
clip creation.

kgarwoodsdzwa committed:
display rms and mel spectrogram together

displays 2 charts with the same timestep, the top one is the rms
for each of your specified frame lengths and the bottom is a
mel spectrogram. you can see the correlation between the two
and it can be helpful

kgarwoodsdzwa committed:
add tools to run multiple files

beginnings of way to run multiple files through this process.
needs to create different filenames to differentiate the original
wavs still but then it should be gravy

kgarwoodsdzwa committed:
add outfile and fix filenames

filenames were assuming 1 wav file before, now it writes the name
of the clip with the original wav file in the filename so its
clearer, it also successfully usilizes the proper specified
outpath to save the files

kgarwoodsdzwa committed:
added docstring fix pylint

there is one error of catching too broad an except but im going
to ignore it for right now because it helps with debugging

kgarwoodsdzwa committed:
add yaml file example

kgarwoodsdzwa committed:
fixed linting and a logic error

the last catch for making the final segment in a wav file was
creating a new right_index to stop the recording, when it should
have just been using the stop_index so that it wouldn't go out of
bounds. also made a sub function to make the rms array to clean
up the code a bit

kgarwoodsdzwa committed:
change the function name from main

i want to ultimately put this in the whoot package so it shouldn't
necessarily be called main

kgarwoodsdzwa committed:
fix function name

kgarwoodsdzwa committed:
Merge pull request #77 from conservationtechlab/dev

Dev

kgarwoodsdzwa committed:
reorganize with updated dev to make it be in the whoot package

kgarwoodsdzwa committed:
add some hard coded variables to config

adding the frame_length, the hop_length, and desired clip size to
the config file to easily adjust the values

kgarwoodsdzwa committed:
lint

kgarwoodsdzwa committed:
remove unneeded line

kgarwoodsdzwa committed:
fix the script because import wasnt wrking

valerieporter commented on cfgs/extract_noise_example.yaml:
```diff
@@ -0,0 +1,5 @@
+out: /path/to/output/folder/
+frame_length: 4096
+hop_length: 2048
+num_sec_slice: 3
```
Is num_sec_slice notation slice length or slice duration? Might be good to add a comment explaining this variable.

kgarwoodsdzwa replied:
its how many seconds youd like the resulting extracted clip segments to be

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,39 @@
+        try:
+            print(f"running {file}")
+            clip_loud_segments(os.path.join(config['audio'], file), config)
+        except Exception as e:
```
pylint error: Catching too general exception Exception (broad-exception-caught)

valerieporter commented on tools/extract_noise.py:
```diff
@@ -0,0 +1,108 @@
+import soundfile as sf
+
+
+def clip_loud_segments(file, config):
```
pylint error: Too many local variables (22/15) (too-many-locals).

kgarwoodsdzwa replied:
i usually ignore this specific error, so ill add a pylint ignore comment.

valerieporter commented on tools/extract_noise.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def clip_loud_segments(file, config):
+    """Extract loud segments from a wav file.
```
Maybe add that you are saving this output to a given file path?

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,39 @@
+    for file in all_files:
+        try:
+            print(f"running {file}")
+            clip_loud_segments(os.path.join(config['audio'], file), config)
```
Food for thought: it might be nice to include a return string indicating that the process was successful and where the file was saved?

kgarwoodsdzwa replied:
it does already, located in extract_noise.py in clip_loud_segments. i can however include a line here that counts how many total clips it extracted from an entire audio file, but the code doesn't exit here per clip so that print statement will need to stay in clip_loud_segments

valerieporter commented on tools/display_rms_and_mel.py:
```diff
@@ -0,0 +1,47 @@
+ax[1].set(title='log Power spectrogram')
+
+
+plt.show()
```
Might be nice to add an optional save feature for the graph and text output. (like a True/False variable that is checked by an if statement or something like that?)

valerieporter commented on tools/display_rms_and_mel.py:
```diff
@@ -0,0 +1,47 @@
+FILENAME = '<path/to/audio/file.wav>'
+FRAME_LENGTH = 2048
+HOP_LENGTH = 512
+NUM_SECONDS_OF_SLICE = 3
```
Might be good to add a comment here stating what NUM_SECONDS_OF_SLICE is.

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,39 @@
+
+    python3 run_extract_noise.py
+    -config /path/to/extract_noise_copy.yaml
+
```
Would be good to add a "Raises" section, for ExceptionTypes in the code.

valerieporter commented on tools/README.md:
```diff
@@ -0,0 +1,14 @@
+To investigate and understand your raw audio data better, 
+and to be able to isolate potentially significant acoustic
+events to reduce time labeling.
+
```
Might be good to add a description for extract_noise.py, since these functions can be utilized in other scripts.

valerieporter reviewed (changes requested):
## Feedback ##
- See comments for all files, including .yaml and README.md
### tools/display_rms_and_mel.py: ###
  - pylint: no issues, Rating: 10/10
  - flake8: no issues
### tools/extract_noise.py: ###
  - pylint: line 14 - too many local variables, Rating: 9.84/10
  - flake8: no issues
### tools/run_extract_noise.py: ###
  - pylint:  line 38 - broad exception caught, Rating: 9.41/10
  - flake8: no issues

kgarwoodsdzwa committed:
adding descriptions to config values

kgarwoodsdzwa committed:
add total clips saved and make error more specific

kgarwoodsdzwa committed:
add raises to docstring for exception

kgarwoodsdzwa committed:
add better description

kgarwoodsdzwa committed:
disable pylint too many locals and unused e

kgarwoodsdzwa committed:
flake8

kgarwoodsdzwa committed:
fix pylint

kgarwoodsdzwa committed:
add description for extract_noise

kgarwoodsdzwa committed:
save out plot

valerieporter commented on tools/extract_noise.py:
```diff
@@ -0,0 +1,123 @@
+"""
+import os
+import librosa
+import librosa.display
```
Doesn't look like this is being utilized in the code anymore

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,38 @@
+    with open(ARGS.config, 'r', encoding='UTF-8') as f:
+        config = yaml.safe_load(f)
+    all_files = os.listdir(config['audio'])
+    for file in all_files:
```
So I am getting an error that if I have any other file types or folders where my data is stored (I am nesting the output file folder inside the input file folder; input = dir/folder, output = dir/folder/output_folder), then the code fails because its inputting a directory not a file.

Is there a directory structure specified somewhere in whoot? May be that should be specified or maybe there needs to be a file type search add to the all_files line on line 33? I see you have an exception for a file type issue in extract_noise.py on lines 41-45, but maybe all you need is something like this for run_extract_noise.py:

In .yaml file:
`# audio file type for code to look for`
`file_type: '.wav'`

In code (replacing all_files:
`all_files = glob.glob(config['audio']+"*"+file_type)`

Rather than using os.listdir()? You might not need that exception anymore if you incorporate this. Just a thought.

kgarwoodsdzwa replied:
added a check for wavs, (can be uppper or lower case,) and worked for me when i nested the output library but itd be interesting to see if it works for you. still need the error check in the function later in case the file is corrupted. hopefully it addressed what youre referring to!

kgarwoodsdzwa committed:
checks for wavs only so doesnt get confused by nested dirs

kgarwoodsdzwa committed:
remove librosa.display not used

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,39 @@
+    ARGS = PARSER.parse_args()
+    with open(ARGS.config, 'r', encoding='UTF-8') as f:
+        config = yaml.safe_load(f)
+    all_wav_files = [file for file in os.listdir(config['audio'])
```
I wonder if we could include something that will look through multiple file folders within a data folder? i.e. bat_data --> contains 9 species, each with their own folder.

valerieporter reviewed (changes requested):
Works for me now. I have added one more comment about recursively looking through multiple folders inside the give audio dataset folder/path.

kgarwoodsdzwa committed:
make functions out of duplicate code and allow recursivity

Now the function descends into entire directory to obtain the
wavs that could be inside other sub dirs, and writes a metadata
file that correlates the resulting clip with the original file,
along with the relative start time that the clip originated
from in the original file.

kgarwoodsdzwa commented:
it now recursively goes through folders to look for all audio in the provided dir- however, i still have it store all output to one folder. this also means we need a metadata file to associate the clip back to the file it came from, so it generates a metadata file with the original file, the new segment file, and the relative start time of the clip (relative to the original file it came from). I also made each segment be a uuid string as we don't need all the info we had before in that filename, since we have a metadata file. this could also be easier for us in the future if we had labelers look at the segments, we would want the metadata included in the filename likely scrubbed anyways.

valerieporter commented on tools/extract_noise.py:
```diff
@@ -0,0 +1,156 @@
+    segment_id = uuid.uuid4()
+    name = config['out'] + str(segment_id) + ".wav"
+    sound_slice = sound[left_index:right_index]
+    sf.write(name, sound_slice, sr)
```
Bug: If I give it an output folder name (config['out']) that I want to use but does not currently exist, this save/write code fails. Should we add an input and output exist directory check (and maybe add a os.mkdir())?

valerieporter commented on tools/run_extract_noise.py:
```diff
@@ -0,0 +1,46 @@
+                        help='Path to config.')
+    ARGS = PARSER.parse_args()
+    with open(ARGS.config, 'r', encoding='UTF-8') as f:
+        config = yaml.safe_load(f)
```
Maybe the directory exist check can go here? (see other comment for context)

kgarwoodsdzwa replied:
added!

kgarwoodsdzwa committed:
add check to make output directory if it doesnt exist

valerieporter commented on tools/display_rms_and_mel.py:
```diff
@@ -0,0 +1,54 @@
+
+if SAVE_PLOT is True:
+    name = Path(FILENAME).stem
+    plot_name = name + "_RMS_plot.png"
```
plot_name does not have a file folder destination added to it. It saved this image to folder that whoot is contained in instead. Could you save the figure to the folder that audio file is in?

valerieporter commented on tools/extract_noise.py:
```diff
@@ -0,0 +1,156 @@
+                real_index = mid_index * hop_length + int(frame_length/2)
+                half_slice_width = int(num_sec_slice * sr / 2)
+                left_index = max(0, real_index - half_slice_width)
+                if left_index > last_right_index:
```
I believe this is where we are defining the left and right indexes of the rms clip. If so, we may want another if statement checking the length of the remaining clip beyond left_index=num_sec_slice, before proceeding.  

If the length is fine as long as its <num_sec_slice, then disregard this comment. 

But, if this is not the correct outcome and we need the clip to match the length of the other clips, then I think adding zero padding to the end of the audio clip could be useful to extend the clip to the intended length.  If zero padding is not optimal for ML networks for bird/soundscape data (<20kHz), I wonder if adding pink noise (rather than no noise) to the end would be sufficient? 

Could potentially create an additional setting that the user has to input for the type of padded data?

kgarwoodsdzwa replied:
i think its best to not pad at all, and if the user wants equal length clips they can decide how to pad by post processing, the original goal with this was to be able to manually validate the clips, and for that it doesnt need to have padding. i think if we decide we want that as a feature later, we can make an issue to add it then, but for now i dont think its a needed feature

kgarwoodsdzwa committed:
store result plot in same place as wav file
